src/utils/event-sourcing/src/aggregate.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use super::errors::*;
+use crate::EventID;
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+/// Common set of operations for an event-sourced aggregate
+#[async_trait::async_trait]
+#[allow(drop_bounds)]
+pub trait Aggregate
```

Comment:
The heart of event sourcing - I'm using this trait to enforce same structure on all aggregates.

There exist multiple event-sourcing libraries for Rust, but I decided to start with my own implementation to get a feel for how it can look like in Rust and understand key requirements.

I will then use experience from writing this to evaluate these libraries to hopefully switch to something off-the-shelf.

Reply:
Yes you are right. I was going to try to refactor this into `Aggregate<Projection>` where `Projection` specifies the state transition logic and `Aggregate` type implements all the event handling. I'd prefer to do this in a separate PR though, after I finish API server integration. Don't want to be stuck perfecting the event sourcing "framework" without progressing to the actual goal which is to have dataset updates running in the platform backend.